src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -357,6 +358,159 @@ private static byte[] UrlEncode(byte[] bytes, int offset, int count)
             return expandedBytes;
         }
 
+        private static unsafe string UrlEncode(char* s, int length)
+        {
+            int numUnsafe = 0;
+            int numSpaces = 0;
+            for (int i = 0; i < length; i++)
+            {
+                char c = s[i];
+                if (c == ' ')
+                {
+                    numSpaces++;
+                }
+                else if (!IsUrlSafeChar(s[i]))
+                {
+                    numUnsafe += GetByteCount(s, i, length);
```

Comment:
This implementation is considerably more efficient than Encoding.UtF8's implementation according to my benchmarks. Therefore, this custom implementation, and the one below are partly responsible for the increase in performance in cases where any kind of escaping is required